### PR TITLE
Fix url parse

### LINKED ISSUE #144: Unable to log in with provided credentials. - Non-Alphanumeric Characters
I am trying to log in with the API, but I am no longer able to. 

rh = Robinhood()
logged_in = rh.login(username='', password='')
print(logged_in)

My username does have NON alphanumeric characters--is this what is causing the issue?

What solutions are there for me?

### PR DESCRIPTION
**Pull request by m-deck**
closes #144. need some good review on this, there were a few things i was considering changing

### CODE REVIEW

**Comment by nkulung** [comment]
Is Jamonek still active to approve the merge? Or should I make changes locally

**Comment by m-deck** [comment]
@nkulung I'm not sure when this might be approved but I've been happily using this change locally for the last week. If you have the repo cloned, just check out this branch and then use `pip install -e`.

**Comment by nkulung** [comment]
@m-deck Thanks

**Comment by aamazie** [comment]
@m-deck Your code doesn't fix the whole/actual problem. See Issue #147

File: Robinhood/Robinhood.py [hunk 1]
```diff
@@ -100,32 +102,44 @@ def login(self,
...
-        if 'token' in data.keys():
-            self.auth_token = data['token']
-            self.headers['Authorization'] = 'Token ' + self.auth_token
-            return True
+            data = urllib.urlencode(fields) #py2
+        except:
+            #data = urllib.parse.urlencode(fields) #py3
+            data = fields
+
+        print(data)
```
Comment (stoooops) [inline Robinhood/Robinhood.py]: Delete this line. It will print your password to sysout

**Comment by briansrls** [comment]
@aamazie @stoooops I believe it works for users without 2fa.  I tried disabling mine and it actually worked, earlier I was getting the  {'detail': 'Authentication credentials were not provided.'} error.  I don't recommend turning off 2fa unless you're desperate (I've already re-enabled mine).

File: Robinhood/Robinhood.py [hunk 2]
```diff
@@ -100,32 +102,44 @@ def login(self,
...
-        self.password = password
-        payload = {
-            'password': self.password,
-            'username': self.username
-        }
 
         if mfa_code:
-            payload['mfa_code'] = mfa_code
-
+            fields = { 'client_id' : 'c82SH0WZOsabOXGP2sxqcj34FxkvfnWRZBKlBjFS',
```
Comment (vargasa) [inline Robinhood/Robinhood.py]: Regarding `client_id`. Check [this comment](https://github.com/Jamonek/Robinhood/pull/150#discussion_r232499886) out

File: Robinhood/Robinhood.py [hunk 3]
```diff
@@ -100,32 +102,44 @@ def login(self,
...
+                        'expires_in' : 86400,
+                        'grant_type': 'password',
+                        'password' : self.password,
+                        'scope' : 'internal',
+                        'username' : self.username,
+                        'mfa_code': self.mfa_code,
+                       }
+        else:
+            fields = {'client_id': 'c82SH0WZOsabOXGP2sxqcj34FxkvfnWRZBKlBjFS',
```
Comment (vargasa) [inline Robinhood/Robinhood.py]: Same here

File: Robinhood/Robinhood.py [hunk 4]
```diff
@@ -100,32 +102,44 @@ def login(self,
...
+        #print(data)
+        try:
+            #self.auth_token = res['token']
+            self.oauth_token = res['access_token']
+        except KeyError:
+            return res
 
-        return False
+        #self.headers['Authorization'] = 'Token '+self.auth_token
+        self.headers['Authorization'] = 'Bearer ' + self.oauth_token
```
Comment (vargasa) [inline Robinhood/Robinhood.py]: As I understand this `Bearer ` is not fixed but is also retrieved from auth endpoint as a token type

File: Robinhood/Robinhood.py [hunk 5]
```diff
@@ -100,32 +102,44 @@ def login(self,
...
+            #data = urllib.parse.urlencode(fields) #py3
+            data = fields
+
+        print(data)
+        res = self.session.post(endpoints.login(), data=data)
+        #res.raise_for_status()
+        res = res.json()
+        #print(data)
+        try:
+            #self.auth_token = res['token']
```
Comment (vargasa) [inline Robinhood/Robinhood.py]: If commented code is not gonna be used, take it out

File: Robinhood/endpoints.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
 def login():
-    return "https://api.robinhood.com/api-token-auth/"
+    return "https://api.robinhood.com/oauth2/token/"
```
Comment (vargasa) [inline Robinhood/endpoints.py]: `https://api.robinhood.com/` should be defined as a variable it`self`

**Comment by aamazie** [comment]
@vargasa If I don't add a description to my commit, it's either because it would have been redundant from the last commit (i.e. I prpbably mishandled something that I thought I had handled and had to correct it) or I'm getting rid of something minor like an old comment that doesn't make sense. My repo is most likely the most up-to-date and functional. I built on top of @anthonykrivonos 's work with the submit order functions by making a submit_buy_order and submit_sell_order that defaults to an ask/bid price respectively. I also deprecated the place_order function and its derivative functions, as they weren't working anymore. @anthonykrivonos and @thescienceofsciences killed it with solving the authentication issue. I also had fixed the cancel_order function back in May. Everything's now working properly in my repo as far as I know.

**Comment by vargasa** [comment]
> @vargasa <https://github.com/vargasa> If I don't add a description to my 
> commit, it's either because it would have been redundant from the last 
> commit (i.e. I prpbably mishandled something that I thought I had 
> handled and had to correct it) or I'm getting rid of something minor 
> like an old comment that doesn't make sense. My repo is most likely the 

Even in that case you can use 'commit --amend' and 'rebase'

**Pull request closed without merging**